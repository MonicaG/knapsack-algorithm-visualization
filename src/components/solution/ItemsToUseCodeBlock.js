import React from 'react';
import { KnapsackAlgorithmPropType } from './helpers/PropTypesHelper'
import PropTypes from 'prop-types';
import { solutionTableActionTypes as types } from './SolutionController';
import { buildSyntaxHighlight } from './helpers/CodeBlocksCommon';
import ItemsToUseCode from '../../models/codeblock/ItemsToUseCode';
import ControlButtons from './ControlButtons';
import { TITLE_STEP_3, STEP_BTN_NAME, step3Instructions } from './helpers/constants';

function ItemsToUseCodeBlock({ knapsackAlgorithm, state, dispatch }) {

  const codeBlock = new ItemsToUseCode(state.solutionIndex, state.currentCapacity, knapsackAlgorithm);
  return (
    <div>
      <ControlButtons buttonName={STEP_BTN_NAME} buttonAction={handleButtonClick}/>
      <div className="py-2">
        {buildSyntaxHighlight(codeBlock.getCode(), codeBlock.isInSolutions(), codeBlock.getInLineNums(), codeBlock.getOutLineNums())}
      </div>
      <div className="explanation">
        <p>It is now time to find the items that fit in the knapsack. This step checks if each item fits in the knapsack, starting with the last item. It starts there because that item's last cell contains the maximum value of the knapsack.</p>
        <p>Click the <span className="italic">{STEP_BTN_NAME}</span> button to find the items. A green cell means the item fits in the knapsack for the given capacity. A grey cell means the item did  not fit.</p>
        <p>For each item, the algorithm checks if the current cell matches the previous cell's value. If it does NOT match, it means the item is part of the solution. Why? In step 1 of the algorithm the previous cell's value was copied when:</p>
        <ul className="list-disc ml-8">
          <li>the item's weight was greater than the knapsack's capacity, or</li>
          <li>the previous cell's value is greater than the item's value plus the value at any remaining capacity </li>
        </ul>
        <p>A different value means a new, greater value was calculated for the capacity. So, the item is part of the solution.</p>
        <p>If the item is part of the solution, then the capacity is set to the current capacity minus this item's weight. Why? The item may not take up the entire knapsack's capacity.  One or more items may fit in the left-over capacity. So, the next item uses the left-over capacity to see if it fits.</p>
        <p>Note: the above code snippet uses an else statement to show the case when the cell values match. The else block is here for demonstration purposes. Omit it in an actual implementation.</p>
      </div>
    </div>
  )

  function doDispatch(solutionIndex, capacity) {
    dispatch({
      type: types.STEP_FIND_NEXT_SOLUTION_ITEM,
      currentCapacity: capacity,
      solutionIndex: solutionIndex,
      title: TITLE_STEP_3,
      instructions: step3Instructions(STEP_BTN_NAME),
    });
  }

  function handleButtonClick() {
    const solutionIndex = state.solutionIndex - 1;
    let currentCapacity = state.currentCapacity;
    if (knapsackAlgorithm.solutionTable[state.solutionIndex][state.currentCapacity] !== knapsackAlgorithm.solutionTable[state.solutionIndex - 1][state.currentCapacity]) {
      currentCapacity -= knapsackAlgorithm.items[state.solutionIndex - 1].weight
    }
    doDispatch(solutionIndex, currentCapacity);
  }
}

ItemsToUseCodeBlock.propTypes = {
  knapsackAlgorithm: KnapsackAlgorithmPropType.isRequired,
  state: PropTypes.shape({
    solutionIndex: PropTypes.number.isRequired,
    currentCapacity: PropTypes.number.isRequired,
  }).isRequired,
  dispatch: PropTypes.func.isRequired
};

export default ItemsToUseCodeBlock;