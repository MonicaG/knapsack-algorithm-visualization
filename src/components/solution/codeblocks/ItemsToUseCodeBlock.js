import PropTypes from 'prop-types';
import {ItemPropType, KnapsackAlgorithmPropType} from '../helpers/PropTypesHelper'
import {buildSyntaxHighlight} from './CodeBlocksCommon';
import {format} from '../helpers/Formatting';

function ItemsToUseCodeBlock({ previousItem, index, currentCapacity, knapsackAlgorithm }) {
 
  const isSolutionItem = knapsackAlgorithm.solutionTable[index][currentCapacity] !== knapsackAlgorithm.solutionTable[index-1][currentCapacity];
  const inSolutionLineNums = [7,8,9,10]
  const noOpLineNums = [11]

  function getCode() {
    return (
`// items: array of item objects. An item has a value and weight property
let solution = []; 
let capacity = knapsack.capacity;
for (let i = items.length; i > 0; i--) {
  // i: index (${index})
  if (T[i][capacity] != Table[i-1][capacity]) { // T[${index}][${currentCapacity}] != T[${index - 1}][${currentCapacity}]
    //item is part of the solution
    solution.push(item);
    capacity = capacity - items[i-1].weight;  ${!isSolutionItem ? "": 
    `
    //capacity = ${currentCapacity} - ${previousItem.weight} = ${format(currentCapacity - previousItem.weight)}`
  }
  }else {
    // no - op
  }
}`)
  }

  
  return (
    buildSyntaxHighlight(getCode(), isSolutionItem, inSolutionLineNums, noOpLineNums)
  )
}

ItemsToUseCodeBlock.propTypes = {
  previousItem: ItemPropType.isRequired,
  index: PropTypes.number.isRequired,
  currentCapacity: PropTypes.number.isRequired,
  knapsackAlgorithm: KnapsackAlgorithmPropType.isRequired
};

export default ItemsToUseCodeBlock;