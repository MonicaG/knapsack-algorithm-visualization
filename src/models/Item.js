import { nanoid } from 'nanoid'

class Item {

  constructor(name, value, weight) {
    this.id = nanoid();
    this.name = name;
    this.value = value;
    this.weight = weight
  }

  toString() {
    return `Item  id: ${this.id} name: ${this.name} value: ${this.value} weight: ${this.weight}`;
  }
}

export default Item;