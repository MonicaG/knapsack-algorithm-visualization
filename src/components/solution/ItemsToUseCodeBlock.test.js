import renderer from 'react-test-renderer';
import KnapSackAlgorithm from '../../models/KnapsackAlgorithm';
import Item from '../../models/Item';
import ItemsToUseCodeBlock from './ItemsToUseCodeBlock';

/** 
 * Using snapshot testing because the syntax hightlighting uses a lot of span tags. This makes it difficult to match text as the text 
 * is spread across multiple elements. 
*/

describe('the code block', () => {
  const item1 = new Item('item1', 4, 2);
  const item2 = new Item('item2', 3, 1);
  const item3 = new Item('item3', 5, 3);
  const items = [item1, item2, item3];
  const capacity = 5

  const algorithm = new KnapSackAlgorithm(items, capacity);
  const mockDispatch = jest.fn();

  it('displays when an item is part of the solution', () => {
    /**
     * The test is setup for the solution to be at the first step, which is the last item, max capacity.
     * Item 3 is part of the solution
     */
     const state = {
      solutionIndex: 3,
      currentCapacity: capacity
    }
    const tree = renderer
      .create(<ItemsToUseCodeBlock
        knapsackAlgorithm = {algorithm}
        state = {state} 
        dispatch = {mockDispatch} />)
      .toJSON();

    expect(tree).toMatchSnapshot()
  });

  it('displays when an item is not part of the solution', () => {
     /**
     * The test is setup for the solution to be at the second step, which is index 2. The capacity is 2, because we are building off
     * of step 1. There is only 2 weight units left in the knapsack.
     * Item 2 is NOT part of the solution even though it fits, because the previous item provides a higher value.
     */
      const state = {
        solutionIndex: 2,
        currentCapacity: capacity - item3.weight
      }
    const tree = renderer
      .create(<ItemsToUseCodeBlock
        knapsackAlgorithm = {algorithm}
        state = {state} 
        dispatch = {mockDispatch} />)
      .toJSON();

    expect(tree).toMatchSnapshot()
  });

});