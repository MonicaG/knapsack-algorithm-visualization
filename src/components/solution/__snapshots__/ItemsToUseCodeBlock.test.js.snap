// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the code block displays when an item is not part of the solution 1`] = `
<div>
  <div
    className="flex justify-between mt-4"
  >
    <input
      className="btnReset"
      onClick={[Function]}
      type="button"
      value="Reset"
    />
    <input
      className="btnBlue"
      onClick={[Function]}
      type="button"
      value="Step"
    />
  </div>
  <div
    className="py-2"
  >
    <pre
      style={
        Object {
          "background": "#f8f8ff",
          "color": "#000",
          "display": "flex",
          "overflowX": "auto",
          "padding": "0.5em",
        }
      }
    >
      <code
        className="language-javascript"
        style={
          Object {
            "whiteSpace": "pre",
          }
        }
      >
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            1
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // items: array of item objects. An item has a value and weight property
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            2
          </span>
          <span
            style={Object {}}
          />
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            let
          </span>
          <span
            style={Object {}}
          >
             solution = []; 

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            3
          </span>
          <span
            style={Object {}}
          />
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            let
          </span>
          <span
            style={Object {}}
          >
             capacity = knapsack.capacity;

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            4
          </span>
          <span
            style={Object {}}
          />
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            for
          </span>
          <span
            style={Object {}}
          >
             (
          </span>
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            let
          </span>
          <span
            style={Object {}}
          >
             i = items.length; i &gt; 
          </span>
          <span
            style={
              Object {
                "color": "#40a070",
              }
            }
          >
            0
          </span>
          <span
            style={Object {}}
          >
            ; i--) {

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            5
          </span>
          <span
            style={Object {}}
          >
              
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // i: index (2)
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            6
          </span>
          <span
            style={Object {}}
          >
              
          </span>
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            if
          </span>
          <span
            style={Object {}}
          >
             (T[i][capacity] != Table[i-
          </span>
          <span
            style={
              Object {
                "color": "#40a070",
              }
            }
          >
            1
          </span>
          <span
            style={Object {}}
          >
            ][capacity]) { 
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // T[2][2] != T[1][2]
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            7
          </span>
          <span
            style={Object {}}
          >
                
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            //item is part of the solution
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            8
          </span>
              solution.push(item);

        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            9
          </span>
          <span
            style={Object {}}
          >
                capacity = capacity - items[i-
          </span>
          <span
            style={
              Object {
                "color": "#40a070",
              }
            }
          >
            1
          </span>
          <span
            style={Object {}}
          >
            ].weight;  

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            10
          </span>
          <span
            style={Object {}}
          >
              }
          </span>
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            else
          </span>
          <span
            style={Object {}}
          >
             {

          </span>
        </span>
        <span
          style={
            Object {
              "backgroundColor": "#BEF264",
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            11
          </span>
          <span
            style={Object {}}
          >
                
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // no - op
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            12
          </span>
            }

        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            13
          </span>
          }
        </span>
      </code>
    </pre>
  </div>
  <div
    className="explanation"
  >
    <p>
      It is now time to find the items that fit in the knapsack. This step checks if each item fits in the knapsack, starting with the last item. It starts there because that item's last cell contains the maximum value of the knapsack.
    </p>
    <p>
      Click the 
      <span
        className="italic"
      >
        Step
      </span>
       button to find the items. A green cell means the item fits in the knapsack for the given capacity. A grey cell means the item did  not fit.
    </p>
    <p>
      For each item, the algorithm checks if the current cell matches the previous cell's value. If it does NOT match, it means the item is part of the solution. Why? In step 1 of the algorithm the previous cell's value was copied when:
    </p>
    <ul
      className="list-disc ml-8"
    >
      <li>
        the item's weight was greater than the knapsack's capacity, or
      </li>
      <li>
        the previous cell's value is greater than the item's value plus the value at any remaining capacity 
      </li>
    </ul>
    <p>
      A different value means a new, greater value was calculated for the capacity. So, the item is part of the solution.
    </p>
    <p>
      If the item is part of the solution, then the capacity is set to the current capacity minus this item's weight. Why? The item may not take up the entire knapsack's capacity.  One or more items may fit in the left-over capacity. So, the next item uses the left-over capacity to see if it fits.
    </p>
    <p>
      Note: the above code snippet uses an else statement to show the case when the cell values match. The else block is here for demonstration purposes. Omit it in an actual implementation.
    </p>
  </div>
</div>
`;

exports[`the code block displays when an item is part of the solution 1`] = `
<div>
  <div
    className="flex justify-between mt-4"
  >
    <input
      className="btnReset"
      onClick={[Function]}
      type="button"
      value="Reset"
    />
    <input
      className="btnBlue"
      onClick={[Function]}
      type="button"
      value="Step"
    />
  </div>
  <div
    className="py-2"
  >
    <pre
      style={
        Object {
          "background": "#f8f8ff",
          "color": "#000",
          "display": "flex",
          "overflowX": "auto",
          "padding": "0.5em",
        }
      }
    >
      <code
        className="language-javascript"
        style={
          Object {
            "whiteSpace": "pre",
          }
        }
      >
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            1
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // items: array of item objects. An item has a value and weight property
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            2
          </span>
          <span
            style={Object {}}
          />
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            let
          </span>
          <span
            style={Object {}}
          >
             solution = []; 

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            3
          </span>
          <span
            style={Object {}}
          />
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            let
          </span>
          <span
            style={Object {}}
          >
             capacity = knapsack.capacity;

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            4
          </span>
          <span
            style={Object {}}
          />
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            for
          </span>
          <span
            style={Object {}}
          >
             (
          </span>
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            let
          </span>
          <span
            style={Object {}}
          >
             i = items.length; i &gt; 
          </span>
          <span
            style={
              Object {
                "color": "#40a070",
              }
            }
          >
            0
          </span>
          <span
            style={Object {}}
          >
            ; i--) {

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            5
          </span>
          <span
            style={Object {}}
          >
              
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // i: index (3)
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            6
          </span>
          <span
            style={Object {}}
          >
              
          </span>
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            if
          </span>
          <span
            style={Object {}}
          >
             (T[i][capacity] != Table[i-
          </span>
          <span
            style={
              Object {
                "color": "#40a070",
              }
            }
          >
            1
          </span>
          <span
            style={Object {}}
          >
            ][capacity]) { 
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // T[3][5] != T[2][5]
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "backgroundColor": "#BEF264",
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            7
          </span>
          <span
            style={Object {}}
          >
                
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            //item is part of the solution
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "backgroundColor": "#BEF264",
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            8
          </span>
              solution.push(item);

        </span>
        <span
          style={
            Object {
              "backgroundColor": "#BEF264",
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            9
          </span>
          <span
            style={Object {}}
          >
                capacity = capacity - items[i-
          </span>
          <span
            style={
              Object {
                "color": "#40a070",
              }
            }
          >
            1
          </span>
          <span
            style={Object {}}
          >
            ].weight;  

          </span>
        </span>
        <span
          style={
            Object {
              "backgroundColor": "#BEF264",
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            10
          </span>
          <span
            style={Object {}}
          >
                
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            //capacity = 5 - 3 = 2
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            11
          </span>
          <span
            style={Object {}}
          >
              }
          </span>
          <span
            style={
              Object {
                "color": "#954121",
              }
            }
          >
            else
          </span>
          <span
            style={Object {}}
          >
             {

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            12
          </span>
          <span
            style={Object {}}
          >
                
          </span>
          <span
            style={
              Object {
                "color": "#408080",
                "fontStyle": "italic",
              }
            }
          >
            // no - op
          </span>
          <span
            style={Object {}}
          >
            

          </span>
        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            13
          </span>
            }

        </span>
        <span
          style={
            Object {
              "display": "block",
            }
          }
        >
          <span
            className="comment linenumber react-syntax-highlighter-line-number"
            style={
              Object {
                "display": "inline-block",
                "minWidth": "2.25em",
                "paddingRight": "1em",
                "textAlign": "right",
                "userSelect": "none",
              }
            }
          >
            14
          </span>
          }
        </span>
      </code>
    </pre>
  </div>
  <div
    className="explanation"
  >
    <p>
      It is now time to find the items that fit in the knapsack. This step checks if each item fits in the knapsack, starting with the last item. It starts there because that item's last cell contains the maximum value of the knapsack.
    </p>
    <p>
      Click the 
      <span
        className="italic"
      >
        Step
      </span>
       button to find the items. A green cell means the item fits in the knapsack for the given capacity. A grey cell means the item did  not fit.
    </p>
    <p>
      For each item, the algorithm checks if the current cell matches the previous cell's value. If it does NOT match, it means the item is part of the solution. Why? In step 1 of the algorithm the previous cell's value was copied when:
    </p>
    <ul
      className="list-disc ml-8"
    >
      <li>
        the item's weight was greater than the knapsack's capacity, or
      </li>
      <li>
        the previous cell's value is greater than the item's value plus the value at any remaining capacity 
      </li>
    </ul>
    <p>
      A different value means a new, greater value was calculated for the capacity. So, the item is part of the solution.
    </p>
    <p>
      If the item is part of the solution, then the capacity is set to the current capacity minus this item's weight. Why? The item may not take up the entire knapsack's capacity.  One or more items may fit in the left-over capacity. So, the next item uses the left-over capacity to see if it fits.
    </p>
    <p>
      Note: the above code snippet uses an else statement to show the case when the cell values match. The else block is here for demonstration purposes. Omit it in an actual implementation.
    </p>
  </div>
</div>
`;
